extra_data/caldata.py

```diff
@@ -0,0 +1,1016 @@
+
+    def set_pdu_ccv(self, modno, calibration_name, entry):
+        """Set a PDU CCV entry.
+
+        Convenience method to handle insertion of keys if missing.
+        """
+
+        pdu_metadata = self.get(modno, dict())
+        pdu_metadata[calibration_name] = entry
+        self[modno] = pdu_metadata
+
+    def __str__(self):
+        """Pretty-print CCV metadata using pandas."""
+
+        import pandas as pd
+
+        res = {pdu_idx: {cal_name: ccv_data['ccv_name']
+                         for cal_name, ccv_data in pdu_data.items()}
+               for pdu_idx, pdu_data in self.items()}
+
+        return str(pd.DataFrame.from_dict(res, orient='index'))
+
+
+class CalCatError(RuntimeError):
+    """CalCat API error."""
+
+    def __init__(self, response):
+        super().__init__(response['info'])
+
+
+class CalibrationData:
+    """Calibration data for detectors.
+
+    European XFEL uses a web app and database to store records about the
+    characterization of detectors and the data necessary to their
+    correction, collectively called CalCat. The default installation is
+    available at https://in.xfel.eu/calibration.
+
+    A detector is identified by a name (e.g. SPB_DET_AGIPD1M-1) and
+    consists of one or more detector modules. The modules are a virtual
+    concept and may be identified by their number (e.g. 3), the Karabo
+    data aggregator in EuXFEL's DAQ system they're connected to
+    (e.g. AGIPD05) or a virtual device name describing their relative
+    location (e.g. Q3M2).
+
+    A detector module is mapped to an actual physical detector unit
+    (PDU), which may be changed in case of a physical replacement. When
+    characterization data is inserted into the database, it is attached
+    to the PDU currently mapped to a module and not the virtual module
+    itself.
+
+    Characterization data is organized by its type just called
+    calibration (e.g. Offset or SlopesFF) and the operating condition it
+    was taken in, which is a mapping of parameter keys to their values
+    (e.g. Sensor bias voltage or integration time). Any unique
+    combination of calibration (type) and operating condition is a
+    calibration constant (CC). Any individual measurement of a CC is
+    called a calibration constant version (CCV). There may be many CCVs
+    for any given CC.
+
+    Note that while a connection to CalCat is possible from anywhere,
+    the actual calibration data referred to is only available on the
+    European XFEL infrastructure.
+    """
+
+    get_detector_keys = ['id', 'name', 'identifier', 'karabo_name',
+                         'karabo_id_control', 'description']
+    get_pdu_keys = ['id', 'physical_name', 'karabo_da', 'virtual_device_name',
+                    'detector_type_id', 'detector_id', 'description']
+
+    calibrations = set()
+
+    client = None
+
+    def __init__(self, detector_name, modules=None, client=None,
+                 **condition_params):
```

Comment:
If the `.from_data()` classmethod is meant as the primary public API to construct these classes, I might have `__init__` accept `condition_params` as a dict rather than keyword arguments. Not a big deal, but having fewer places taking arbitrary keyword args makes life easier if you want to add other specific arguments later on.